20231214-listUsers-api.md

```diff
@@ -0,0 +1,889 @@
+
+## How it Works
+[how-it-works]: #how-it-works
+
+## Algorithm
+
+`ListUsers` and `StreamedListUsers` is a filtered form of recursive [Expand](https://openfga.dev/api/service#/Relationship%20Queries/Expand). We don’t want to return *all users of any type*, we only want to return all the concrete user objects *of a specified type*.
+
+The implementation of ListUsers and StreamedListUsers will behave a lot like [ListObjects](https://openfga.dev/api/service#/Relationship%20Queries/ListObjects) and [StreamedListObjects](https://openfga.dev/api/service#/Relationship%20Queries/StreamedListObjects), but instead of starting at a user and expanding backwards (e.g. reverse expansion) we will start at a relationship with an object and recursively expand (forward expansion) any usersets we find along the way that would lead to a concrete/terminal object matching one of the specific user filters.
+
+
+> ℹ️ Most of the implementation details we spent time tuning with `ListObjects` apply to this problem as well. Namely, bounding the number of concurrent evaluation paths, bounding the number of concurrent database queries that can be inflight per request, constraining the response results and/or the streaming deadline, etc.. We should be able to move more quickly on the implementation phase as a result of the prior work and education from ListObjects.
+
+Here’s are a few examples to demonstrate the algorithm:
+
+### Example 1a (Direct Relation with Object)
+```
+model
+  schema 1.1
+
+type user
+
+type document
+  relations
+    define viewer: [user]
+```
+
+| object     | relation | user        |
+|------------|----------|-------------|
+| document:1 | viewer   | user:jon    |
+| document:1 | viewer   | user:andres |
+
+
+```
+ListUsers({
+  object: "document:1",
+  relation: "viewer",
+  user_filters: [{type: "user"}]
+}) --> {
+  users: ["user:jon", "user:andres"],
+  wildcard_types: []
+}
+```
+In this case the algorithm simply must expand all direct relationships between `document:1#viewer` and `user` objects. Since these are directly related to one another, then we must only do a simple reverse database lookup.
+
+### Example 1b (Direct Relation with Userset)
+```
+model
+  schema 1.1
+
+type user
+
+type group
+  relations
+    define member: [user, group#member]
+
+type document
+  relations
+    define viewer: [group#member]
+```
+| object         | relation | user                  |
+|----------------|----------|-----------------------|
+| document:1     | viewer   | group:eng#member      |
+| group:eng      | member   | group:fga#member      |
+| group:fga      | member   | user:andres           |
+| group:fga      | member   | group:fga-core#member |
+| group:fga-core | member   | user:jon              |
+
+```
+ListUsers({
+  object: "document:1",
+  relation: "viewer",
+  user_filters: [{type: "user"}]
+}) --> ["user:jon", "user:andres"]
```

Comment:
We should stick to the same response format for all examples
```
--> {
  users: ["user:jon", "user:andres"],
  wildcard_types: []
}
```

Reply:
Fixed.